adoc/chapters/programming_interface.adoc

```diff
@@ -288,271 +288,410 @@ context doesn't match the target backend.
+  host application.
+  If [code]#T# is a valid kernel argument, [code]#T# must also be copy
+  constructible and copy assignable within SYCL kernel functions.
+  Any instance of [code]#T# that is a copy of another instance must behave as-if
+  it were the original instance and must represent the same underlying
+  <<native-backend-object>>, where applicable.
+  Any action performed on such a copy must behave as-if it were also performed
+  on the original instance.
+
+* _Destructibility:_ [code]#T# must be destructible in the host application.
+  If [code]#T# is a valid kernel argument, [code]#T# must also be destructible
+  within SYCL kernel functions.
+  When an instance of [code]#T# is destroyed, any copies of the [code]#T#
+  instance (as defined above) are unaffected.
+  When the last remaining host copy of an instance of [code]#T# is destroyed,
+  lifetime-related operations may be performed on the underlying
+  implementation-defined resource, including, where applicable, the destruction
+  of a <<native-backend-object>>.
+
+* _Movability:_ [code]#T# must be move constructible and move assignable in the
+  host application.
+  If [code]#T# is a valid kernel argument, [code]#T# must also be move
+  constructible and move assignable within SYCL kernel functions.
+  Any instance of [code]#T# that is the result of a move from another instance
+  must behave as-if it were the original instance and must represent the same
+  underlying <<native-backend-object>>, where applicable.
+  Move construction and move assignment render the moved-from instance of
+  [code]#T# invalid.
+
+* _Equality:_ [code]#T# must be equality comparable in the host application.
+  If [code]#T# is a valid kernel argument, [code]#T# must also be equality
+  comparablewithin SYCL kernel functions.
+  Equality between two instances of [code]#T# (i.e., [code]#a == b#) must be
+  true if one instance is a copy of the other, and non-equality between two
+  instances of [code]#T# (i.e., [code]#a != b#) must be true if neither instance
+  is a copy of the other, in accordance with the above definition of a copy.
+  An instance of [code]#T# that has been invalidated by a move operation is not
+  considered to be a copy of any other [code]#T# instance.
+  Equality on [code]#T# must be guaranteed to be reflexive (i.e., [code]#a ==
+  a#), symmetric (i.e., [code]#a == b# implies [code]#b == a# and [code]#a != b#
+  implies [code]#b != a#), and transitive (i.e., [code]#a == b && b == c#
+  implies [code]#c == a#).
+
+* _Hashability:_ A specialization of [code]#std::hash# for [code]#T# must exist
+  in the host application.
+  The specialization of [code]#std::hash# must return a unique value such that
+  for two instances of [code]#T#, [code]#a# and [code]#b#, either [code]#(a ==
+  b) && std::hash(a) == std::hash(b)# evaluates to true, or [code]#(a != b) &&
+  std::hash(a) != std::hash(b)# evaluates to true.
```

Comment:
The text didn't originally say that `operator==()` and `hash()` are available on device. Should we add those?